tsl/src/remote/dist_txn.c

```diff
@@ -474,11 +476,27 @@ dist_txn_xact_callback_2pc(XactEvent event, void *arg)
 static void
 dist_txn_xact_callback(XactEvent event, void *arg)
 {
+	bool use_2pc;
+
 	/* Quick exit if no connections were touched in this transaction. */
 	if (store == NULL)
 		return;
 
-	if (ts_guc_enable_2pc)
+	/*
+	 * The decision to use 2PC rests on multiple factors:
+	 *
+	 * 1) if ts_guc_enable_2pc is enabled and it's a regular backend use it
+	 *
+	 * 2) if (1) is true but only ONE DN is involved in the remote transaction, don't use it
+	 *
+	 * 3) if ts_guc_enable_2pc is enabled but we are running a read only txn, don't use it
+	 */
+	use_2pc = (ts_guc_enable_2pc && !XactReadOnly && (hash_get_num_entries(store->hashtable) != 1));
```

Comment:
I don't think 1 DN doing 1PC is safe. For example:

- AN pre-commit: remotely commits DN
- AN rolls-back transaction due to conflicts
- No way for DN to rollback because no 2PC